Add COSE support to Rekor

kommendorkapten opened the pull request:
#### Summary
Basic support for COSE envelopes.
Added is functionality to upload COSE records, optionally signed with AAD. The AAD is not stored in Rekor.
The COSE envelope is stored as an attestation if attestation storage is enabled, and can so be retrieved via Rekor. The returned envelope is returned in the `attestation` field, not within the `body` field. 

What's stored is the hash of the payload and the entire envelope. If the content type of the COSE envelope is an in-toto statement, the payload is parsed and the subject hashes are extracted and indexed for searching.

#### Ticket Link
This pull requests completes https://github.com/sigstore/rekor/pull/731.

#### Release Note



```release-note
Added new request/response type: COSE envelopes. When uploading a COSE envelope, only the hash of the payload and the hash of the envelope is stored, unless attestation storage is enabled, then the entire envelope is stored as an attestation.
A new command line parameter `--aad` is added. This parameter provides extra Additional Authenticated Data to be used during COSE signature verification (using AAD with COSE is optional).
```

dlorenc committed:
WIP: Add COSE support to Rekor

This commit adds COSE Sign1 support to rekor via a new data type.
COSE is defined in RFC8152, and provides a signing message envelope.

This is supported in rekor using the veraison/go-cose library. The new
API type requires the signed content, the signature envelope, and the public key.

The public key is in the standard rekor PKI format, at the moment only ECDSA P256
with SHA256 is supported. The signed message only supports in-line bodies (no URL fetching),
and there is no support for pre-hashed entries.

Signed-off-by: Dan Lorenc <lorenc.d@gmail.com>

kommendorkapten committed:
Completed basic support for COSE records.

This adds some more functionality related to COSE enveleopes.
Features added are:

 - Support for specifying Additional Authentincated Data (AAD)

 - The entire CBOR envelope is stored as an attestation

 - If the payload type is an in-toto statement, subject is indexed

What's not optimal is that the COSE envelope is using the regular
`Attestion()` functionality, which means that rekor cli tries to
print it during `rekor-cli get` and the response record from Rekor
looks a bit awkward.

Signed-off-by: Fredrik Skogman <kommendorkapten@github.com>

kommendorkapten committed:
Updated the documentation for COSE envelopes.

Signed-off-by: Fredrik Skogman <kommendorkapten@github.com>

bobcallaway commented on cmd/rekor-cli/app/pflag_groups.go:
```diff
@@ -87,6 +87,11 @@ func addArtifactPFlags(cmd *cobra.Command) error {
 		},
+		"aad": {
+			base64Flag,
+			"base64 encoded aad",
```
you may want to spell out what `aad` means here, since this will be printed as part of command line flag help output.

bobcallaway commented on pkg/types/cose/cose.go:
```diff
@@ -0,0 +1,73 @@
+//
+// Copyright 2021 The Sigstore Authors.
```
can you switch to 2022 throughout all new files please?

kommendorkapten replied:
Of course, my bad!

kommendorkapten committed:
Resolved merge conflicts with main.

The biggest change is adapting the new interface where attestation func
is split to two, one to get the key and a nother to get key/val.

Signed-off-by: Fredrik Skogman <kommendorkapten@github.com>

kommendorkapten committed:
Ran go mod tidy after resolving merge committs.

Signed-off-by: Fredrik Skogman <kommendorkapten@github.com>

dlorenc commented on pkg/types/cose/v0.0.1/entry.go:
```diff
@@ -0,0 +1,326 @@
+	switch t := cryptoPub.(type) {
+	case *rsa.PublicKey:
+		alg = gocose.AlgorithmPS256
+	case *ecdsa.PublicKey:
```
Do we need to be a bit more careful here to avoid curve mismatches? Is there a possibility someone actually used PS384 or PS512?

kommendorkapten replied:
Yeah, good catch. I can play around a bit more early next week to see if we can tighten up this.

dlorenc commented on pkg/types/cose/v0.0.1/entry.go:
```diff
@@ -0,0 +1,326 @@
+	}
+
+	// If payload is an in-toto statement, let's grab the subjects.
+	if rawContentType, ok := v.sign1Msg.Headers.Protected[gocose.HeaderLabelContentType]; ok {
```
This part is interesting, I'm not sure I've seen in-toto statements embedded into COSE envelopes before. I think it makes sense overall though.

dlorenc reviewed (commented):
One main question on algorithm checking.

kommendorkapten committed:
Added check to see that provided EC key uses the P256 curve.

Signed-off-by: Fredrik Skogman <kommendorkapten@github.com>

bobcallaway commented on go.mod:
```diff
@@ -25,6 +25,7 @@ require (
 	github.com/mediocregopher/radix/v4 v4.1.0
 	github.com/mitchellh/go-homedir v1.1.0
 	github.com/mitchellh/mapstructure v1.5.0
+	github.com/pkg/errors v0.9.1
```
there was a recent effort to stop using this and use `errors` instead.

bobcallaway commented on pkg/types/cose/cose.go:
```diff
@@ -0,0 +1,73 @@
+import (
+	"context"
+
+	"github.com/pkg/errors"
```
can you swap this with `"errors"` instead?

bobcallaway commented on pkg/types/cose/v0.0.1/entry.go:
```diff
@@ -0,0 +1,350 @@
+
+	alg, pk, err := getPublicKey(v.keyObj)
+	if err != nil {
+		return nil
```
did you mean to return `err` here?

kommendorkapten replied:
Of course, will go over some test coverage here to make sure we got it covered.

bobcallaway commented on pkg/types/cose/v0.0.1/entry.go:
```diff
@@ -0,0 +1,350 @@
+	if err := v.sign1Msg.UnmarshalCBOR(v.CoseObj.Message); err != nil {
+		return err
+	}
+
```
should we ensure unmarshaling is successful before assigning the output to `v.sign1Msg`?

kommendorkapten replied:
You mean like this:
```go
        coseMsg := gocose.NewSign1Message()
	if err := coseMsg.UnmarshalCBOR(v.CoseObj.Message); err != nil {
		return err
	}
        v.sign1Msg = coseMsg
```

I think that makes sense, we can even wait until the signature is verified until we assign it to `v.sign1Msg`.

bobcallaway replied:
yes, ideally I wouldn't want to partially change the pointer receiver unless all validation is successful.

bobcallaway commented on pkg/types/entries.go:
```diff
@@ -121,12 +121,13 @@ func CanonicalizeEntry(ctx context.Context, entry EntryImpl) ([]byte, error) {
-	PublicKeyPath  *url.URL
-	PublicKeyBytes []byte
-	PKIFormat      string
+	AdditionalAuthenticatedData string
```
should this get changed to `[]byte` and leave it up to the caller to decode the input correctly?

kommendorkapten replied:
Yes, that makes sense! So basically moving the base64 decoding into the cli part instead of `CreteFromArtifactProperties`. I'll give it a shot to see if I can make it happen.

bobcallaway reviewed (commented):
looking good overall, just a few questions/nits

kommendorkapten committed:
Spelled out aad when printing the help message.

Signed-off-by: Fredrik Skogman <kommendorkapten@github.com>

kommendorkapten committed:
Updated copyright notice to have current (2022) year.

Signed-off-by: Fredrik Skogman <kommendorkapten@github.com>

kommendorkapten committed:
Removed direct dependency on github.com/pkg/errors and replaced with stdlib
errors package.

Signed-off-by: Fredrik Skogman <kommendorkapten@github.com>

kommendorkapten committed:
Fixed a bug where nil was wrongfully returned instead of err.
Also increas the general test coverage.

Signed-off-by: Fredrik Skogman <kommendorkapten@github.com>

kommendorkapten committed:
Changed aad in artifact properties struct to be []byte instead of string.
This gives the caller the possibility to decide how to decode the data.

Signed-off-by: Fredrik Skogman <kommendorkapten@github.com>

bobcallaway commented on cmd/rekor-cli/app/pflags_test.go:
```diff
@@ -346,6 +347,23 @@ func TestArtifactPFlags(t *testing.T) {
 			expectParseSuccess:    true,
 			expectValidateSuccess: false,
 		},
+		{
```
maybe one more test case for invalid base64 data passed on the command line?

bobcallaway reviewed (commented):
LGTM, would like one more test case to ensure we have coverage for non-base64 encoded input on the `aad` cmd line argument

kommendorkapten committed:
Fixed a warning from the linter.

Signed-off-by: Fredrik Skogman <kommendorkapten@github.com>

kommendorkapten committed:
Added test case for malformed base64 aad parameter.

Signed-off-by: Fredrik Skogman <kommendorkapten@github.com>

kommendorkapten committed:
During signature validation, do not store any state until entire validation
is done.

Signed-off-by: Fredrik Skogman <kommendorkapten@github.com>
